core/src/erasure.rs

```diff
@@ -1,234 +1,152 @@
-// Support erasure coding
-use crate::packet::{Blob, SharedBlob};
-use crate::result::{Error, Result};
+use crate::packet::{Blob, SharedBlob, BLOB_HEADER_SIZE};
+use crate::result::Result;
 use std::cmp;
+use std::convert::AsMut;
 use std::sync::{Arc, RwLock};
 
+use reed_solomon_erasure::ReedSolomon;
+
 //TODO(sakridge) pick these values
-pub const NUM_DATA: usize = 16; // number of data blobs
-pub const NUM_CODING: usize = 4; // number of coding blobs, also the maximum number that can go missing
-pub const ERASURE_SET_SIZE: usize = NUM_DATA + NUM_CODING; // total number of blobs in an erasure set, includes data and coding blobs
-
-macro_rules! align {
-    ($x:expr, $align:expr) => {
-        $x + ($align - 1) & !($align - 1)
-    };
-}
+/// Number of data blobs
+pub const NUM_DATA: usize = 16;
+/// Number of coding blobs; also the maximum number that can go missing.
+pub const NUM_CODING: usize = 4;
+/// Total number of blobs in an erasure set; includes data and coding blobs
+pub const ERASURE_SET_SIZE: usize = NUM_DATA + NUM_CODING;
 
-#[derive(Debug, PartialEq, Eq)]
-pub enum ErasureError {
-    NotEnoughBlocksToDecode,
-    DecodeError,
-    EncodeError,
-    InvalidBlockSize,
-    InvalidBlobData,
-    CorruptCoding,
-}
+pub fn encode(data: &[&[u8]], parity: &mut [&mut [u8]]) -> Result<()> {
+    let rs = ReedSolomon::new(data.len(), parity.len())?;
+    rs.encode_sep(data, parity)?;
 
-// k = number of data devices
-// m = number of coding devices
-// w = word size
-
-extern "C" {
-    fn jerasure_matrix_encode(
-        k: i32,
-        m: i32,
-        w: i32,
-        matrix: *const i32,
-        data_ptrs: *const *const u8,
-        coding_ptrs: *const *mut u8,
-        size: i32,
-    );
-    fn jerasure_matrix_decode(
-        k: i32,
-        m: i32,
-        w: i32,
-        matrix: *const i32,
-        row_k_ones: i32,
-        erasures: *const i32,
-        data_ptrs: *const *mut u8,
-        coding_ptrs: *const *mut u8,
-        size: i32,
-    ) -> i32;
-    fn galois_single_divide(a: i32, b: i32, w: i32) -> i32;
-    fn galois_init_default_field(w: i32) -> i32;
+    Ok(())
 }
 
-use std::sync::Once;
-static ERASURE_W_ONCE: Once = Once::new();
+/// Recover data + coding blocks into data blocks
+/// # Arguments
+/// * `data` - array of data blocks to recover into
+/// * `coding` - array of coding blocks
+/// * `erasures` - list of indices in data where blocks should be recovered
+pub fn decode_blocks(
+    blocks: &mut [&mut [u8]],
+    erasures: &[usize],
+    data_count: usize,
+    coding_count: usize,
+) -> Result<()> {
+    let rs = ReedSolomon::new(data_count, coding_count)?;
```

Comment:
is this worth hanging onto, similar to the generator?

Reply:
Looking at the `ReedSolomon` struct's definition I think so actually, since it contains a fair amount of data, it's silly to reconstruct over and over again.

I'll make `&ReedSolomon` the first argument of all these methods and then store an `Arc` with one in `Blocktree` and pass it along to the coding generator.

Reply:
or make a wrapper Recoverer or Decoder that matches Generator and wraps up the ReedSolomon instance in order to match the types coming out of BlockTree

Reply:
another option is a lazy_static!()  ;)